internal/slicer/slicer.go

```diff
@@ -220,96 +238,37 @@ func Run(options *RunOptions) (*Report, error) {
+			if len(pathInfo.Arch) > 0 && !slices.Contains(pathInfo.Arch, arch) {
 				continue
 			}
 			if done[relPath] || pathInfo.Kind == setup.CopyPath || pathInfo.Kind == setup.GlobPath {
 				continue
 			}
 			done[relPath] = true
-			addKnownPath(relPath)
-			targetPath := filepath.Join(targetDir, relPath)
-			targetMode := pathInfo.Mode
-			if targetMode == 0 {
-				if pathInfo.Kind == setup.DirPath {
-					targetMode = 0755
-				} else {
-					targetMode = 0644
-				}
-			}
-
-			// Leverage tar handling of mode bits.
-			tarHeader := tar.Header{Mode: int64(targetMode)}
-			var fileContent io.Reader
-			var linkTarget string
-			switch pathInfo.Kind {
-			case setup.TextPath:
-				tarHeader.Typeflag = tar.TypeReg
-				fileContent = bytes.NewBufferString(pathInfo.Info)
-			case setup.DirPath:
-				tarHeader.Typeflag = tar.TypeDir
-			case setup.SymlinkPath:
-				tarHeader.Typeflag = tar.TypeSymlink
-				linkTarget = pathInfo.Info
-			default:
-				return nil, fmt.Errorf("internal error: cannot extract path of kind %q", pathInfo.Kind)
+			data := pathData{
+				until:   pathInfo.Until,
+				mutable: pathInfo.Mutable,
 			}
-
-			entry, err := fsutil.Create(&fsutil.CreateOptions{
-				Path:        targetPath,
-				Mode:        tarHeader.FileInfo().Mode(),
-				Data:        fileContent,
-				Link:        linkTarget,
-				MakeParents: true,
-			})
+			addKnownPath(knownPaths, relPath, data)
+			targetPath := filepath.Join(targetDir, relPath)
+			entry, err := createFile(targetPath, pathInfo)
 			if err != nil {
 				return nil, err
 			}
 			err = report.Add(slice, entry)
 			if err != nil {
 				return nil, err
 			}
-			if pathInfo.Until == setup.UntilMutate {
-				pathUntil[relPath] = setup.UntilMutate
-			}
 		}
 	}
 
 	// Run mutation scripts. Order is fundamental here as
 	// dependencies must run before dependents.
-	checkWrite := func(path string) error {
-		if !pathInfos[path].Mutable {
-			return fmt.Errorf("cannot write file which is not mutable: %s", path)
-		}
-		return nil
-	}
-	checkRead := func(path string) error {
-		var err error
-		if !knownPaths[path] {
-			// we assume that path is clean and ends with slash if it designates a directory
-			if path[len(path)-1] == '/' {
-				if path == "/" {
-					panic("internal error: content root (\"/\") is not selected")
-				}
-				if knownPaths[path[:len(path)-1]] {
-					err = fmt.Errorf("content is not a directory: %s", path[:len(path)-1])
-				} else {
-					err = fmt.Errorf("cannot list directory which is not selected: %s", path)
-				}
-			} else {
-				if knownPaths[path+"/"] {
-					err = fmt.Errorf("content is not a file: %s", path)
-				} else {
-					err = fmt.Errorf("cannot read file which is not selected: %s", path)
-				}
-			}
-		}
-		return err
-	}
+	checker := contentChecker{knownPaths}
 	content := &scripts.ContentValue{
 		RootDir:    targetDirAbs,
-		CheckWrite: checkWrite,
-		CheckRead:  checkRead,
+		CheckWrite: checker.checkMutable,
+		CheckRead:  checker.checkKnown,
```

Comment:
Not sure we want to break away from the original names, but we can roll with that for now and see how far we get.